olp-cpp-sdk-dataservice-write/src/IndexLayerClientImpl.cpp

```diff
@@ -153,101 +154,63 @@ client::CancellationToken IndexLayerClientImpl::PublishIndex(
     return client::CancellationToken();
   }
 
-  auto op_id = tokenList_.GetNextId();
-  auto cancel_context = std::make_shared<client::CancellationContext>();
-  auto self = shared_from_this();
-  auto cancel_function = [=]() {
-    self->tokenList_.RemoveTask(op_id);
-    callback(PublishIndexResponse(client::ApiError(
-        client::ErrorCode::Cancelled, "Operation cancelled.", true)));
-  };
-
   const auto data_handle = GenerateUuid();
 
-  auto insert_indexes_callback =
-      [=](InsertIndexesResponse insert_indexes_response) {
-        self->tokenList_.RemoveTask(op_id);
-        if (!insert_indexes_response.IsSuccessful()) {
-          callback(PublishIndexResponse(insert_indexes_response.GetError()));
-          return;
-        }
-        model::ResponseOkSingle res;
-        res.SetTraceID(data_handle);
-        callback(PublishIndexResponse(std::move(res)));
-      };
-
-  auto insert_indexes_function = [=]() -> client::CancellationToken {
-    auto index = request.GetIndex();
-    index.SetId(data_handle);
-    return IndexApi::insertIndexes(
-        *self->apiclient_index_, index, request.GetLayerId(),
-        request.GetBillingTag(), insert_indexes_callback);
-  };
+  auto publish_task =
```

Comment:
Please be aware that you also need to change L145 to L155 as both are calling the user callback sync which is not very good. All should be triggered on the TaskScheduler.